README.md
# WSUWP YouTube Embed

Use the `[wsuwp_youtube]` shortcode to embed YouTube videos on a site using the [YouTube iFrame Player API](https://developers.google.com/youtube/iframe_api_reference).

## Shortcode attributes

* `video_id` _ID of the YouTube video._
* `width` **560** _Width in pixels of the video embed._
* `height` **315** _Height in pixels of the video embed._
* `autoplay` **0** _If set to `1`, autoplay the video on load._
* `controls` **1** _If set to `0`, don't show additional video controls._
* `loop` **0** _If set to `1`, loop the video automatically._
* `modestbranding` **1** _If set to `0`, show default YouTube branding._
* `playsinline` **0** _If set to `1`, play videos inline in iOS._
* `rel` **0** _If set to `1`, show related videos when play stops._
* `showinfo` **0** _If set to `1`, show information about the video in the embed._
* `start` **0** _Number of seconds into the video at which it should start._
* `end` _Number of seconds into the video at which it should end._
* `volume` **default** _Accepts "mute" or an integer between 0 and 100._


CHANGELOG.md
# Changelog

## 0.0.2

* Fix inner wrapper class name typo.

## 0.0.1

* Initial plugin release.
* Supported attributes:
    * `video_id`
    * `width`
    * `height`
    * `autoplay`
    * `controls`
    * `loop`
    * `modestbranding`
    * `playsinline`
    * `rel`
    * `showinfo`
    * `start`
    * `end`
    * `volume`


tests/bootstrap.php
<?php

$_tests_dir = getenv( 'WP_TESTS_DIR' );
if ( ! $_tests_dir ) {
	$_tests_dir = '/tmp/wordpress-tests-lib';
}

require_once $_tests_dir . '/includes/functions.php';

function _manually_load_plugin() {
	require dirname( __FILE__ ) . '/../wsuwp-youtube-embed.php';
}
tests_add_filter( 'muplugins_loaded', '_manually_load_plugin' );

require $_tests_dir . '/includes/bootstrap.php';



tests/test-wsuwp-youtube-embed.php
<?php

class WSUWP_YouTube_Embed_Tests extends WP_UnitTestCase {
	/**
	 * An initial sample test to verify working tests.
	 */
	public function test_wsuwp_youtube_embed_function() {
		$this->assertTrue( true );
	}
}


js/video.js
/* global YT, onPlayerReady */
( function( $, window ) {
	/**
	 * Create a script element to load in the YouTube iFrame API and insert it
	 * into the document.
	 */
	var load_youtube = function() {
		var tag = document.createElement( "script" );

		tag.src = "https://www.youtube.com/iframe_api";
		var firstScriptTag = document.getElementsByTagName( "script" )[ 0 ];
		firstScriptTag.parentNode.insertBefore( tag, firstScriptTag );
	};

	/**
	 * Callback function expected by the YouTube Iframe API. Without a function
	 * with this name available in the global space, our use of the YouTube API
	 * does not work.
	 *
	 * Loop through each of the inline YouTube Videos, gather the video information,
	 * and set up objects representing the videos.
	 */
	window.onYouTubeIframeAPIReady = function() {
		$( ".wsuwp-youtube-embed" ).each( function() {
			var video_id = $( this ).data( "video-id" ),
				video_height = $( this ).data( "video-height" ),
				video_width = $( this ).data( "video-width" ),
				video_autoplay = $( this ).data( "video-autoplay" ),
				video_controls = $( this ).data( "video-control" ),
				video_end = $( this ).data( "video-end" ),
				video_loop = $( this ).data( "video-loop" ),
				video_modestbranding = $( this ).data( "video-modestbranding" ),
				video_playsinline = $( this ).data( "video-playsinline" ),
				video_rel = $( this ).data( "video-rel" ),
				video_showinfo = $( this ).data( "video-showinfo" ),
				video_start = $( this ).data( "video-start" );

			new YT.Player( "wsuwp-youtube-video-" + video_id, {
				height: video_height,
				width: video_width,
				videoId: video_id,
				playerVars: {
					autoplay: video_autoplay,
					controls: video_controls,
					end: video_end,
					loop: video_loop,
					modestbranding: video_modestbranding,
					playsinline: video_playsinline,
					rel: video_rel,
					showinfo: video_showinfo,
					start: video_start
				},
				events: {
					"onReady": onPlayerReady
				}
			} );
		} );
	};

	/**
	 * Handles custom behavior when the player is ready.
	 *
	 * @param event
	 */
	window.onPlayerReady = function( event ) {
		var volume = $( event.target.h ).data( "video-volume" );
		if ( "default" === volume ) {
			return;
		} else if ( "mute" === volume ) {
			event.target.mute();
		} else {
			event.target.setVolume( volume );
		}
	};

	/**
	 * Fire any actions that we need to happen once the document is ready.
	 */
	$( document ).ready( function() {
		load_youtube();
	} );
}( jQuery, window ) );


Gruntfile.js
module.exports = function( grunt ) {
    grunt.initConfig( {
        pkg: grunt.file.readJSON( "package.json" ),

        concat: {
            options: {
                sourceMap: false
            },
            video_js: {
                src: [ "src/js/video.js" ],
                dest: "js/video.js"
            }
        },

        phpcs: {
            plugin: {
                src: [ "./*.php", "./includes/*.php" ]
            },
            options: {
                bin: "vendor/bin/phpcs --extensions=php --ignore=\"*/vendor/*,*/node_modules/*\"",
                standard: "phpcs.ruleset.xml"
            }
        },

        jscs: {
            scripts: {
                src: [ "Gruntfile.js", "src/js/*.js" ],
                options: {
                    preset: "jquery",
                    requireCamelCaseOrUpperCaseIdentifiers: false, // We rely on name_name too much to change them all.
                    maximumLineLength: 250
                }
            }
        },

        jshint: {
            grunt_script: {
                src: [ "Gruntfile.js" ],
                options: {
                    curly: true,
                    eqeqeq: true,
                    noarg: true,
                    quotmark: "double",
                    undef: true,
                    unused: false,
                    node: true     // Define globals available when running in Node.
                }
            },
            theme_scripts: {
                src: [ "src/js/*.js" ],
                options: {
                    bitwise: true,
                    curly: true,
                    eqeqeq: true,
                    forin: true,
                    freeze: true,
                    noarg: true,
                    nonbsp: true,
                    quotmark: "double",
                    undef: true,
                    unused: true,
                    browser: true, // Define globals exposed by modern browsers.
                    jquery: true   // Define globals exposed by jQuery.
                }
            }
        },

        uglify: {
            video_js: {
                src: "js/video.js",
                dest: "js/video.min.js"
            }
        }
    } );

    grunt.loadNpmTasks( "grunt-contrib-concat" );
    grunt.loadNpmTasks( "grunt-phpcs" );
    grunt.loadNpmTasks( "grunt-jscs" );
    grunt.loadNpmTasks( "grunt-contrib-jshint" );
    grunt.loadNpmTasks( "grunt-contrib-uglify" );

    // Default task(s).
    grunt.registerTask( "default", [ "jscs", "jshint", "concat", "uglify", "phpcs" ] );
};


includes/class-wsuwp-youtube-embed.php
<?php

class WSUWP_YouTube_Embed {
	/**
	 * @since 0.1.1
	 *
	 * @var WSUWP_YouTube_Embed
	 */
	private static $instance;

	/**
	 * Tracks the version number of the plugin for script enqueues.
	 *
	 * @since 0.0.1
	 *
	 * @var string
	 */
	public $version = '0.0.2';

	/**
	 * Maintains and returns the one instance. Initiates hooks when
	 * called the first time.
	 *
	 * @since 0.0.1
	 *
	 * @return \WSUWP_YouTube_Embed
	 */
	public static function get_instance() {
		if ( ! isset( self::$instance ) ) {
			self::$instance = new WSUWP_YouTube_Embed();
			self::$instance->setup_hooks();
		}
		return self::$instance;
	}

	/**
	 * Adds hooks.
	 *
	 * @since 0.0.1
	 */
	public function setup_hooks() {
		add_shortcode( 'wsuwp_youtube', array( $this, 'display_wsuwp_youtube_shortcode' ) );
	}

	/**
	 * Displays the HTML associated with the `wsuwp_youtube` shortcode.
	 *
	 * @since 0.0.1
	 *
	 * @param array $atts
	 *
	 * @return string
	 */
	public function display_wsuwp_youtube_shortcode( $atts ) {
		$defaults = array(
			'video_id' => '',
			'width' => '560',
			'height' => '315',
			'autoplay' => 0,
			'controls' => 1,
			'end' => '',
			'loop' => 0,
			'modestbranding' => 1,
			'playsinline' => 0,
			'rel' => 0,
			'showinfo' => 0,
			'start' => 0,
			'volume' => 'default',
		);

		$atts = shortcode_atts( $defaults, $atts );

		array_walk( $atts, array( $this, 'sanitize_atts' ) );

		ob_start();
		?>
		<!-- Embedded with wsuwp_youtube shortcode. -->
		<div class="wsuwp-youtube-wrap-outer">
			<div class="wsuwp-youtube-wrap-inner">
				<div class="wsuwp-youtube-embed"
				     id="wsuwp-youtube-video-<?php echo esc_attr( $atts['video_id'] ); ?>"
				     data-video-id="<?php echo esc_attr( $atts['video_id'] ); ?>"
				     data-video-width="<?php echo absint( $atts['width'] ); ?>"
				     data-video-height="<?php echo absint( $atts['height'] ); ?>"
				     data-video-autoplay="<?php echo absint( $atts['autoplay'] ); ?>"
				     data-video-controls="<?php echo absint( $atts['controls'] ); ?>"
				     data-video-end="<?php echo absint( $atts['end'] ); ?>"
				     data-video-loop="<?php echo absint( $atts['loop'] ); ?>"
				     data-video-modestbranding="<?php echo absint( $atts['modestbranding'] ); ?>"
				     data-video-playsinline="<?php echo absint( $atts['playsinline'] ); ?>"
				     data-video-rel="<?php echo absint( $atts['rel'] ); ?>"
				     data-video-showinfo="<?php echo absint( $atts['showinfo'] ); ?>"
				     data-video-start="<?php echo absint( $atts['start'] ); ?>"
				     data-video-volume="<?php echo esc_attr( $atts['volume'] ); ?>"></div>
			</div>
		</div>
		<!-- End wsuwp_youtube embed. -->
		<?php
		$content = ob_get_contents();
		ob_end_clean();

		wp_enqueue_script( 'wsuwp-youtube-embed', plugins_url( '/js/video.min.js', dirname( __FILE__ ) ), array( 'jquery' ), $this->version, true );

		return $content;
	}

	/**
	 * Sanitizes the attributes passed to the shortcode based on the
	 * expectations of each key.
	 *
	 * @since 0.0.1
	 *
	 * @param string|int $value
	 * @param string     $key
	 */
	public function sanitize_atts( &$value, $key ) {
		if ( 'video_id' === $key ) {
			$value = sanitize_text_field( $value );
		} elseif ( in_array( $key, array( 'width', 'height', 'start', 'end' ), true ) ) {
			$value = absint( $value );
		} elseif ( 'volume' === $key && in_array( $value, array( 'default', 'mute' ), true ) ) {
			return; // value remains the same.
		} else {
			$value = absint( $value );
			if ( 1 < $value ) {
				$value = 0;
			}
		}
	}
}
